mitiq/cdr/clifford_training_data.py

```diff
@@ -0,0 +1,442 @@
+    num_angles: int, random_state: np.random.RandomState
+) -> np.ndarray:
+    """Returns an array of Clifford angles chosen uniformly at random.
+
+    Args:
+        num_angles: Number of Clifford angles to return in array.
+        random_state: Random state for sampling.
+    """
+    return np.array(
+        [random_state.choice(_CLIFFORD_ANGLES) for _ in range(num_angles)]
+    )
+
+
+@np.vectorize
+def _closest_clifford(angles: np.ndarray) -> float:
+    """Returns the nearest Clifford angles to the input angles.
+
+    Args:
+        non_Clifford_ops: Non-Clifford opperations.
+    """
+    ang_scaled = angles / (np.pi / 2)
+    # if just one min value, return the corresponding nearest cliff.
+    if (
+        abs((ang_scaled / 0.5) - 1) > 10 ** (-6)
+        and abs((ang_scaled / 0.5) - 3) > 10 ** (-6)
+        and (abs((ang_scaled / 0.5) - 5) > 10 ** (-6))
+    ):
+        index = int(np.round(ang_scaled)) % 4
+        return _CLIFFORD_ANGLES[index]
+    # If equidistant between two Clifford angles, randomly choose one.
+    else:
+        index_list = [ang_scaled - 0.5, ang_scaled + 0.5]
+        index = int(np.random.choice(index_list))
+        return _CLIFFORD_ANGLES[index]
+
+
+@np.vectorize
+def _is_clifford_angle(angles: np.ndarray, tol: float = 10 ** -5,) -> bool:
+    """Function to check if a given angle is Clifford.
+
+    Args:
+        angles: rotation angle in the Rz gate.
+    """
+    angles = angles % (2 * np.pi)
+    closest_clifford_angle = _closest_clifford(angles)
+    if abs(closest_clifford_angle - angles) < tol:
+        return True
+    return False
+
+
+@np.vectorize
+def _make_probability_distribution(
+    angle: np.ndarray, sigma: float
+) -> np.ndarray:
+    """Returns probability distribution based on distance from angles to
+    Clifford gates.
+
+    Args:
+        angle: angle to form probability distribution.
+
+    Returns:
+        discrete value of probability distribution calculated from
+        exp(-(diff/sigma)^2) where diff is the distance from each angle and the
+        Clifford gates.
+    """
+    s_matrix = cirq.unitary(cirq.S)
+    rz_matrix = cirq.unitary(cirq.rz(angle % (2 * np.pi)))
+    # TODO: Update loop / if.
+    dists = []
+    for exponent in range(4):
+        if exponent == 0:
+            exponent = 4
+        diff = np.linalg.norm(rz_matrix - s_matrix ** exponent)
+        dists.append(np.exp(-((diff / sigma) ** 2)))
+    return dists
+
+
+@np.vectorize
+def _angle_to_probabilities(angle: np.ndarray, sigma: float) -> float:
+    """Returns probability distribution based on distance from angles to
+    Clifford gates.
+
+    Args:
+        angle: angle to form probability distribution.
+
+    Returns:
+        discrete value of probability distribution calculated from
+        exp(-(dist/sigma)^2) where dist = sum(dists) is the
+        sum of distances from each Clifford gate.
+    """
+    dists = _make_probability_distribution(angle, sigma)
+    return sum(dists)
```

Comment:
These two functions don't return actual distrubutions or probabilities. It took me a while to understand why `sum(dists)` is not 1. 
The functions return something similar to a closeness measure: "how close is this angle to a clifford angle"? 
So maybe something like `_angle_to_proximities` and `_angle_to_proximity` could be better names for the two functions. I would not mention probabilities.

Also, I am not convinced why the second function returns `sum(dists)` .   Probably the results would be almost the same, but isn't  `max(dists)` a more natural measure of "proximity" ?